spec/edpm-openstack-service-architecture.md

```diff
@@ -0,0 +1,348 @@
+
+* The `CustomServiceConfig` field in the service CRDs simplifies and unifies
+  the OpenStack service configuration for the human deployer regardless where
+  the service is being deployed (i.e. podified or EDPM).
+
+**Drawbacks**:
+
+* The mechanism to watch DataPlane CRs from service operators and auto create
+  service CRs adds implementation complexity and introduces possible
+  performance impacts that could be avoided if dataplane-operator creates the
+  service CRs.
+
+* The Ready condition of the DataPlaneRole and the DataPlaneNode can only
+  represent the phase -1 and phase 0 readiness of the EDPM node. The phase 1
+  readiness can only be collected by checking the Ready condition of every
+  service CRs related to this EDPM node.
+
+
+### Example
+
+NovaCell/nova-cell1
+```yaml
+apiVersion: nova.openstack.org/v1beta1
+kind: NovaCell
+metadata:
+    name: nova-cell1
+...
+```
+No changes in NovaCell.
+
+OpenstackDataPlaneRole/edpm-compute
+```yaml
+apiVersion: dataplane.openstack.org/v1beta1
+kind: OpenStackDataPlaneRole
+metadata:
+    name: edpm-compute
+    annotations:
+        nova.openstack.org/cell: nova-cell1
+spec:
+    dataPlane: openstack-edpm
+    nodeTemplate:
+       ansiblePort: 22
+...
+```
+The `novaTemplate` is removed from the `nodeTemplate` and a new user specified
+annotation `nova.openstack.org/cell` is added so the human deployer can
+define that this Role describe a set of EDPM nodes that are compute nodes
+connected to cell1.
+
+OpenstackDataPlaneNode/edpm-compute-0
+```yaml
+apiVersion: dataplane.openstack.org/v1beta1
+kind: OpenStackDataPlaneNode
+metadata:
+    name: edpm-compute-0
+spec:
+    ansibleHost: 192.168.122.100
+    hostName: edpm-compute-0
+    node:
+        ansibleSSHPrivateKeySecret: dataplane-ansible-ssh-private-key-secret
+    role: edpm-compute
+...
+```
+The `novaTemplate` is removed from the `node`. As this DataPlaneNode refers
+to a DataPlaneRole that has a `nova.openstack.org/cell` annotation, this
+DataPlaneNode is a compute node connected to cell1.
+
+OpenstackDataPlaneNode/edpm-compute-1
+```yaml
+apiVersion: dataplane.openstack.org/v1beta1
+kind: OpenStackDataPlaneNode
+metadata:
+    name: edpm-compute-1
+    annotations:
+        nova.openstack.org/cell: nova-cell2
+spec:
+    ansibleHost: 192.168.122.101
+    hostName: edpm-compute-1
+    node:
+        ansibleSSHPrivateKeySecret: dataplane-ansible-ssh-private-key-secret
+    role: edpm-compute
+...
+```
+This DataPlaneNode has an `nova.openstack.org/cell` annotation that overrides
+the cell annotation of its role. So this EDPM node is a compute node connected
+to cell2.
+
+NovaExternalCompute=edpm-compute-0
+```yaml
+apiVersion: nova.openstack.org/v1beta1
+kind: NovaExternalCompute
+metadata:
+    name: edpm-compute-0
+spec:
+    dataplaneNodeName: edpm-compute-0
+    cellName: nova-cell1
+    customServiceConfig: |
+    # human deployer can set this to pass through service specific config
```

Comment:
When and where? Patch this CR after it is created? (Yes, I see below).

Does the creation of this by the nova-operator immediately start the ansible jobs? (looks like it, once the node CR is ready per below). If so, I have to patch this after creation, which may re-trigger ansible job runs to update the values I set?

Reply:
so i was wondering about that too. we could have a separate `nova.openstack.org/deploy`
annotation which would be True or False and default to false if not set.

or if we also implement the profile concept it could have a deploy field

either approach would allow you to have the service operator (nova in this case) do everything except deploy so you can examine the configs.

as it stands right now deploy true is our default os if we don't change that then yes the nova operator would deploy nova once the dataplane node ready condition is true.

Reply:
Good points. I only considered the case when the human wants to set debug=true for nova-compute on a single EDPM node as day 2 operation. In that case re-running the ansible is needed anyhow to change the config and restart the service. If the compute node specific config is needed during the first deploy already then that extra config either
* a) needs to come from the profile (see #42) 
* b) we need to model a deploy flag. 
  There is a deploy flag in DataPlaneNode.Spec.DeployStrategy.Deploy so we can simply use that as the source of information to set NovaExternalCompute.Spec.Deploy . I think we don't really need two separately settable flag, one compute specific annotation and one EDPM generic Spec field. I noted this now in the spec.

Reply:
we deffintly need two stpes since we are not going to reconsile until the datapalne node is ready right? unless your suggesting if the dataplen node is set to deploy false we would still generate all the config.

that would work if and only if we do not need to read info from the output of the datapalne operator.

i could see use getting PCI info or ips form it in the future so I'm not sure we can really on the dataPlaneNode.Spec.DeployStrategy for the service operators.

as it stand today we need the two levels for adoption to work properly.